kale/interpret/correlation_analysis.py

```diff
@@ -0,0 +1,352 @@
+
+    for model in bootstrap_models:
+        scaled_errors_pred = model.predict(uncertainties_pred)
+        ax.plot(uncertainties_pred[::-1], scaled_errors_pred[::-1], "-", color="grey", alpha=0.2, zorder=2)
+
+
+def _plot_scatter_points(ax, uncertainties: np.ndarray, scaled_errors: np.ndarray, scatter_color: str) -> None:
+    """Plot scatter points of uncertainties vs errors."""
+    ax.scatter(uncertainties, scaled_errors, marker="o", color=scatter_color, zorder=1, alpha=0.2)
+
+
+def _plot_piecewise_segments(
+    ax,
+    piecewise_model: pwlf.PiecewiseLinFit,
+    uncertainties: np.ndarray,
+    quantile_thresholds: List[float],
+    colors: List[str],
+    num_pred_points: int = 20000,
+) -> None:
+    """Plot piecewise linear segments with different colors."""
+    uncertainties_pred = np.linspace(min(uncertainties), max(uncertainties), num=num_pred_points)
+    scaled_errors_pred = piecewise_model.predict(uncertainties_pred)
+
+    for idx in range(len(quantile_thresholds) + 1):
+        color = colors[idx % len(colors)]
+        min_val, max_val = _get_quantile_bounds(quantile_thresholds, uncertainties, idx)
+
+        plot_indices = [i for i in range(len(uncertainties_pred)) if min_val <= uncertainties_pred[i] < max_val]
+
+        # Shade background to make slices per quantile
+        ax.axvspan(
+            uncertainties_pred[plot_indices][0], uncertainties_pred[plot_indices][-1], facecolor=color, alpha=0.1
+        )
+        ax.plot(uncertainties_pred[plot_indices], scaled_errors_pred[plot_indices], "-", color=color, zorder=3)
+
+
+def _setup_plot_formatting(
+    ax,
+    bin_label_locs: List[float],
+    quantile_thresholds: List[float],
+    correlation_dict: Dict[str, List[float]],
+    font_size: int = 25,
+) -> None:
+    """Setup plot formatting including labels, ticks, and correlation text."""
+    # Set x-axis ticks and labels
+    ax.set_xticks(bin_label_locs)
+    new_labels = [r"$Q_{{{}}}$".format(x + 1) for x in range(len(quantile_thresholds) + 1)]
+    ax.xaxis.set_major_formatter(plt.FixedFormatter(new_labels))
+    ax.yaxis.set_major_formatter(ScalarFormatter())
+
+    # Add correlation text
+    smp_corr, smp_p = correlation_dict["spearman"]
+    p_val = 0.001 if np.round(smp_p, 3) == 0 else np.round(smp_p, 3)
+    bold_text = (
+        r"$\bf{\rho:} $" + r"$\bf{{{}}}$".format(np.round(smp_corr, 2)) + ", p-value < " + r"${{{}}}$".format(p_val)
+    )
+
+    ax.text(
+        0.5,
+        0.9,
+        bold_text,
+        size=font_size,
+        color="black",
+        horizontalalignment="center",
+        verticalalignment="center",
+        transform=ax.transAxes,
+    )
+
+    # Set axis labels and formatting
+    ax.set_xlabel("Uncertainty Quantile", fontsize=font_size)
+    ax.set_ylabel("Error (mm)", fontsize=font_size)
+    ax.tick_params(axis="x", labelsize=font_size)
+    ax.tick_params(axis="y", labelsize=font_size)
+
+
+def analyze_uncertainty_correlation(
```

Comment:
Consider moving public APIs to the top of the file before private functions. It can help users quickly locate the main functions.